mlflow/mlflow
# Migrate Mlflow API request to databricks sdk authentication way and support OAuth

<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/WeichenXu123/mlflow/pull/12011?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12011/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12011
```

</p>
</details>

### Related Issues/PRs


 #xxx

### What changes are proposed in this pull request?

* (1) Using Databricks-SDK API to send rest API request, so that it can support various kinds of authentication including **OAuth**
* (2) Note that Databricks-SDK can't support several authentication ways which are supported by current MLflow, to ensure backward compatibility, I added fallback code and made a few refactoring to making code clear.
* (3) Note that Databricks-SDK can't support reading credential token values, but `get_databricks_env_vars` and `MlflowCredentialContext` requires reading token value, I also keep relevant code of reading credential tokens, and made some refactoring. Note that `get_databricks_env_vars` has been used in some customer code such as Ray-on-Spark to set up MLflow authentication in remote Ray tasks, so we still need to support it.

Note that ecause of (2) and (3), in current stage we still have to keep some of legacy databricks-cli python module code.

### How is this PR tested?


**Manual test with OAuth**
1. Prepare OAuth credentials including "client_id" and "client_secret", see guide in https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html 
2. You can tell MLflow to use OAuth by one of the following ways:

a): Set 3 environmental variables: DATABRICKS_HOST + DATABRICKS_CLIENT_ID + DATABRICKS_CLIENT_SECRET
b): Edit `~/.databrickscfg` file, adding a section for specific profile like:
```
[DEFAULT]
host  = https://e2-demo-field-eng.cloud.databricks.com/
client_id = <your client id>
client_secret = <your client secret>
```

In MLflow application (assuming you are running it in your local machine and it needs to log data to databricks shard), you need to set mlflow tracking URI to either `"databricks"` or `"databricks://<your-profile-name>"`, if no profile name in tracking URI, `DEFAULT` profile is used when reading `~/.databrickscfg` file.

then you can run MLflow testing code like:
```
result = mlflow.create_experiment("/Users/weichen.xu@databricks.com/my-mlflow-exp-12345")
print(result)

result = mlflow.get_experiment("642681409487128")
print(result)
```



### Does this PR require documentation update?


### Release Notes

#### Is this a user-facing change?




#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates

Interface


Language


Integrations




<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:


#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>
